zkevm-circuits/src/witness/block.rs

```diff
@@ -180,20 +179,12 @@ pub fn block_convert(
             .collect(),
         end_block_not_last: step_convert(&block.block_steps.end_block_not_last),
         end_block_last: step_convert(&block.block_steps.end_block_last),
-        bytecodes: block
-            .txs()
-            .iter()
-            .flat_map(|tx| {
-                tx.calls()
-                    .iter()
-                    .map(|call| call.code_hash)
-                    .unique()
-                    .into_iter()
-                    .map(|code_hash| {
-                        let bytecode =
-                            Bytecode::new(code_db.0.get(&code_hash).cloned().unwrap_or_default());
-                        (bytecode.hash, bytecode)
-                    })
+        bytecodes: code_db
+            .0
+            .values()
+            .map(|v| {
+                let bytecode = Bytecode::new(v.clone());
+                (bytecode.hash, bytecode)
```

Comment:
Hi @CPerezz and @han0110, I am sorry that I could not confirm if this change is right.

I Fixed to convert all bytecodes of bus-mapping code DB into circuit Block (not only call context bytecodes) in `block_convert` function. Since it causes external account bytecode is failed to lookup in `bytecode_table` via `bytecode_length` function.

I also found a latest change of `EXTCODECOPY` https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/720#issuecomment-1364631519? It seems that this fix could only populate the necessary `ext_code_hash` in corresponding bus-mapping (as invoking `add_ext_code_hash` function).

Could you help review which fix is suitable? Thanks a lot 🙏

Reply:
If the `CircuitInputBuilder` only fetches the necessary witness (which should be desirable), there will be no unused bytecode stored in `code_db`, so I think your approach is much cleaner and simpler.

Reply:
I see. Thanks.

Reply:
hmm thinking more about this for production env (a prover running aside of operating blockchain), perhaps in the future we'd want to take `code_db` as reference which contains all pre-fetched bytecodes (and bytecodes are cached for future blocks), at that moment we might prefer https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/720#issuecomment-1364631519 more.

@CPerezz any idea which solution we should take now?

Reply:
I'm with @han1001 on this. I think we might want to leave it for now.

Reply:
Then let's keep this in mind and merge the PR (in the future when anyone is going to change the behavior of `CircuitInputBuilder`, the one also needs to change this part to avoid blow up Bytecode circuit).